src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
+
 public:
     void append(const segment_meta& meta) { _col.append(meta); }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> begin() const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.begin());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> end() const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.end());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     find(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.find(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     lower_bound(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.lower_bound(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     upper_bound(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.upper_bound(o()));
     }
 
     std::optional<segment_meta> last_segment() const {
+        if (!_write_buffer.empty()) {
+            auto last_committed_offset = _col.last_committed_offset().value_or(
+              model::offset::min());
+            auto buffer_last_base_offset = _write_buffer.crbegin()->first;
+            // CHECK: base_offset is never equal to committed_offset
+            if (likely(buffer_last_base_offset > last_committed_offset)) {
+                // return last one in _write_buffer
+                return _write_buffer.crbegin()->second;
+            }
+            // fallthrough
+        }
         return _col.last_segment();
     }
 
-    void insert(const segment_meta& m) { _col.append(m); }
+    void insert(const segment_meta& m) {
+        auto [m_it, insertion_took_place] = _write_buffer.insert_or_assign(
+          m.base_offset, m);
+        // m could be a replacement segment for one in the buffer in 2 cases:
+        // 1. !insertion_took_place (assignment):
+        //    - key was already in use, m is a replacement for a segment in the
+        //    buffer
+        // 2. insertion took place at the start of the buffer:
+        //    - m could be a replacement for a segment already in the store, and
+        //    there could be an overlap with the segments at the start of the
+        //    write buffer
+        // TODO could it happen that a segment does not replace an existing one
+        // (different base_offset), but still overlaps with the successors? i.e.
+        // a gap in the data where a previous segment is lost, and the current
+        // one would be a replacement for the previous one?
+        if(_write_buffer.size()>1 &&
```

Comment:
This logic is completed, I guess that we just need to replace an element, the replacement logic in the manifest is handled explicitly, there is a method that first finds all elements that will be replaced and moves them to the `partition_manifest::_replaced` collection. And this component should probably behave just like a `btree_map`.

Reply:
Also, maybe there is no real need to return a fragmented_vector of replaced segments from the `replace_entries` for the same reason.